tp030/README.md

```diff
@@ -0,0 +1,88 @@
+
+## Characters
+For the sake of easier naming, let's define the following three DIDs who will appear throughout this TP.
+1. *Alice* is a DWN owner who will be the *delegator*.
+2. *Alice-ChatApp* is the app on Alice's phone that will be the *delegatee*.
+3. *Bob* is the *counterparty* who wants to receive DWN messages signed by Alice-ChatApp with the understanding that Alice-ChatApp is representing Alice. Bob wants to be confident that Alice has not revoked Alice-ChatApp's grant, so he will keep his eyes peeled.
+
+## Granting and Invocation
+Alice creates a `DelegatesGrant` message with her own DID in the `delegatedBy` field and Alice-ChatApp's DID in the `delegatedTo` field.
+When Alice-ChatApp sends a message to Bob invoking the grant, the `authorization` must contain both Alice-ChatApp's signature AND a reference to the `DelegatesGrant` message. The reference may take one of two forms: Either the entire `DelegatesGrant` message may be listed in the `delegatesGrant` field of the `authorization` OR the CID of the `DelegatesGrant` message may appear in the `delegatesGrantId` field of the `authorization`. The fields `delegatesGrant` and `delegatesGrantId` are mutually exclusive.
+
+## Obtaining and Processing Grants
+When Bob receives the message from Alice-ChatApp, either the message contains the entire grant in the `delegatesGrant` field OR it contains a CID of the grant `DelegatesGrant` with. In the former case, if Bob has never seen the grant before, he will process and store it in his DWN. In the latter case, he must already have the grant in his DWN.
+
+## Revocation
+### Phase 1: `DelegatesCheck`
+When Bob authorizes a message with a `DelegatesGrant`, he will send a `DelegatesCheck` message to Alice's to check if the grant is still active. If the grant is active, Alice will say so. If the grant has been revoked, Alice will send the `DelegatesRevoke` message in the response along with a list of message CIDs to retain. To reduce network calls, Bob will cache the result of `DelegatesCheck`.
+
+This first phase only requires `DelegatesGrant`, `DelegatesRevoke`, and `DelegatesCheck` to exist.
+
+### Phase 2: `DelegatesRegister`
+We build on Phase 1 to further reduce network calls. The first time Bob sees a given `DelegatesGrant`, he will send a `DelegatesRegister` to Alice, adding Bob to a list of entities who want to know when a grant is revoked. If the grant has already been revoked when Bob sends the `DelegatesRegister`, he will receive the `DelegatesRevoke` and retention list in the response. After Bob sends the `DelegatesRegister`, he will not need to ping Alice about revocation again. When Alice wants to revoke Alice-ChatApp's grant, she will get the list of DIDs who sent a `DelegatesRegister` for that grant and send the `DelegatesRevoke` to them. If Bob is not able to receive the `DelegatesRevoke`, he will likely know in advance, leading him to use the Phase 1 approach.
+
+## Message structure
+```ts
+type DelegatesGrantDescriptor = {
```

Comment:
Scenarios requires us to incorporate scoping for the delegates grants, which IMO demonstrates very high similarity in needs with permissions grant, thus a main reason why reuse feels appropriate.